mm2src/mm2_main/src/lp_swap/taker_swap.rs

```diff
@@ -2271,6 +2373,191 @@ pub struct TakerSwapPreparedParams {
+        other_pub: other_taker_coin_htlc_pub.as_slice(),
+        secret_hash: &secret_hash,
+        tx: &taker_payment,
+        search_from_block: taker_coin_start_block,
+        swap_contract_address: &taker_coin_swap_contract_address,
+        swap_unique_data: &unique_data,
+        watcher_reward,
+    };
+    let taker_payment_spend_tx = match swap.taker_coin.search_for_swap_tx_spend_my(search_input).await {
+        Ok(found_swap) => found_swap,
+        Err(e) => return ERR!("Error {} when trying to find taker payment spend", e),
+    };
+
+    match (maker_payment_spend_tx, taker_payment_spend_tx) {
+        (
+            Some(FoundSwapTxSpend::Spent(maker_payment_spend_tx)),
+            Some(FoundSwapTxSpend::Spent(taker_payment_spend_tx)),
+        ) => {
+            swap.taker_coin
+                .validate_watcher_spend(maker_payment_spend_tx.clone())
+                .map_err(|e| e.to_string())?;
+
+            if saved.is_finished() {
+                saved.events.pop();
+            }
+
+            if !saved.taker_payment_spent() {
+                let tx_hash = taker_payment_spend_tx.tx_hash();
+                info!("Taker payment spend tx {:02x}", tx_hash);
+                let tx_ident = TransactionIdentifier {
+                    tx_hex: BytesJson::from(taker_payment_spend_tx.tx_hex()),
+                    tx_hash,
+                };
+                let secret = match swap
+                    .taker_coin
+                    .extract_secret(&secret_hash, &tx_ident.tx_hex, watcher_reward)
+                    .await
+                {
+                    Ok(bytes) => H256Json::from(bytes.as_slice()),
+                    Err(_) => {
+                        return ERR!("Could not extract secret from taker payment spend transaction");
+                    },
+                };
+
+                let event = TakerSwapEvent::TakerPaymentSpent(TakerPaymentSpentData {
+                    transaction: tx_ident,
+                    secret,
+                });
+                let to_save = TakerSavedEvent {
+                    timestamp: now_ms(),
+                    event,
+                };
+                saved.events.push(to_save);
+            }
+
+            let tx_hash = maker_payment_spend_tx.tx_hash();
+            info!("Watcher maker payment spend tx {:02x}", tx_hash);
+            let tx_ident = TransactionIdentifier {
+                tx_hex: BytesJson::from(maker_payment_spend_tx.tx_hex()),
+                tx_hash,
+            };
+
+            let event = TakerSwapEvent::MakerPaymentSpentByWatcher(tx_ident);
+            let to_save = TakerSavedEvent {
+                timestamp: now_ms(),
+                event,
+            };
+            saved.events.push(to_save);
+
+            let mut success_events: Vec<String> = saved.events.iter().map(|e| e.event.to_str()).collect();
+            success_events.push("Finished".into());
+            saved.success_events = success_events;
```

Comment:
Yes this was one thing I wasn't sure of. My first implementation was rewriting a made-up event list at taker restart to fit the events to the `TAKER_USING_WATCHERS_SUCCESS_EVENTS`, to make it appear like the taker completed the swap itself. Then I decided that it is good to know that it wasn't the taker itself, but the watcher who completed the swap. And if the taker has saved some events before failure, e.g. "TakerPaymentSpent", I should keep them with the actual timestamps instead of made-up events. So what I do in the current implementation is to just pop the "Finished" event (if the event is finished), append either `MakerPaymentSpentByWatcher` or `TakerPaymentRefundedByWatcher`, and finally append `"Finished"` again. However, this creates a high number of different success cases and the solution I found was to just change the success events in the swap. Do you have any other idea?

Reply:
You should include any new event in these lists
https://github.com/KomodoPlatform/komodo-defi-framework/blob/4dd54592b12bf0db2f7130344fbcac552c4933bd/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L45-L90
This way GUIs have a list of all success and failure events and that's what they list in the swap screen and show which event happens when it does. You can try a swap in Komodo Wallet web to see this in action.

Reply:
Does this list have to be fixed? Currently we have two different success paths, depending on whether the taker uses watchers there will be a `"WatcherMessageSent"` event or not. Will that be handled?

In my current implementation of taker restart, no matter what the state of the saved events are (including the failure events), I was just appending the events `"TakerPaymentSpend"` and `"MakerPaymentSpentByWatcher"` and I changed the implementation of the `is_success()` function such that if the events include `"MakerPaymentSpentByWatcher"` then it is considered a success even if there were failures.

What I can do is, after checking watcher payments and finding the maker payment spend after restart, removing all failure events and events that do not belong to the success events and build this single list of events:

```
    "Started",
    "Negotiated",
    "TakerFeeSent",
    "TakerPaymentInstructionsReceived",
    "MakerPaymentReceived",
    "MakerPaymentWaitConfirmStarted",
    "MakerPaymentValidatedAndConfirmed",
    "TakerPaymentSent",
    "WatcherMessageSent",
    "TakerPaymentSpent",
    "MakerPaymentSpentByWatcher",
    "Finished",
```

in which case we well end up with three different success paths: taker doesn't send a watcher message and completes the swap, taker sends a watcher message and completes the swap, taker sends a watcher message and a watcher completes the swap. If we don't care whether a watcher has completed the swap or the taker, I can also remove the `"MakerPaymentSpentByWatcher"` event and use the regular `"MakerPaymentSpent"` event, so we can have one less success paths.

How are the failure and refund events shown to the user on GUI? In that case there are several paths and we don't know in advance how the swap will progress.

Reply:
> Does this list have to be fixed?

Yes, they are read from here by the GUIs https://github.com/KomodoPlatform/komodo-defi-framework/blob/4dd54592b12bf0db2f7130344fbcac552c4933bd/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L119-L130
And listed on the swap screen.

> Currently we have two different success paths, depending on whether the taker uses watchers there will be a "WatcherMessageSent" event or not. Will that be handled?

We also have `TAKER_SUCCESS_EVENTS` and `TAKER_USING_WATCHERS_SUCCESS_EVENTS` const lists `TAKER_USING_WATCHERS_SUCCESS_EVENTS` is appended to swap data if the coins support watchers. So this is the list that will be used by GUIs. Not all success events have to be fired, but this list checks if the event is part of success event or not.

Reply:
pub const TAKER_USING_WATCHERS_SUCCESS_EVENTS: [&str; 12] = [
    "Started",
    "Negotiated",
    "TakerFeeSent",
    "TakerPaymentInstructionsReceived",
    "MakerPaymentReceived",
    "MakerPaymentWaitConfirmStarted",
    "MakerPaymentValidatedAndConfirmed",
    "TakerPaymentSent",
    "WatcherMessageSent",
    "TakerPaymentSpent",
    "MakerPaymentSpent",
    "MakerPaymentSpentByWatcher",
    "Finished",
];

Is this what you mean? The list will include both `"MakerPaymentSpent"` and `"MakerPaymentSpentByWatcher"`, but only one of them will be fired.

So do you think I should remove all events that don't belong to the success events, and complete the events according to the success events? For example, on restart we had a swap with the events

```
    "Started",
    "Negotiated",
    "TakerFeeSent",
    "TakerPaymentInstructionsReceived",
    "MakerPaymentReceived",
    "MakerPaymentWaitConfirmStarted",
    "MakerPaymentValidatedAndConfirmed",
    "TakerPaymentSent",
    "WatcherMessageSent",
    "TakerPaymentSpent",
    "MakerPaymentSpendFailed",
    "Finished",
```

I remove the `"Finished"` and `"MakerPaymentSpendFailed"` events, then I add `"MakerPaymentSpentByWatcher"` and `"Finished"` events and sabe the swap as:

```
    "Started",
    "Negotiated",
    "TakerFeeSent",
    "TakerPaymentInstructionsReceived",
    "MakerPaymentReceived",
    "MakerPaymentWaitConfirmStarted",
    "MakerPaymentValidatedAndConfirmed",
    "TakerPaymentSent",
    "WatcherMessageSent",
    "TakerPaymentSpent",
    "MakerPaymentSpentByWatcher",
    "Finished",
```

Reply:
> Is this what you mean? The list will include both "MakerPaymentSpent" and "MakerPaymentSpentByWatcher", but only one of them will be fired.

Yes, the static list of events is unrelated to the actual events fired.

Actual events:
https://github.com/KomodoPlatform/komodo-defi-framework/blob/7ad7cc1069580273f3a383032210fa60350836ad/mm2src/adex_cli/src/rpc_data/swaps.rs#L78

Static List:
https://github.com/KomodoPlatform/komodo-defi-framework/blob/7ad7cc1069580273f3a383032210fa60350836ad/mm2src/adex_cli/src/rpc_data/swaps.rs#L87-L90

Reply:
Okay now I understand, I will make changes to re-order the events and form a single success path on restart.

Reply:
Done

Reply:
Thank you for fixing this after DM discussions :)